src/json.cpp

```diff
@@ -1917,57 +1950,58 @@ void TextJsonIn::error( int offset, const std::string &message )
     err << "\n\n";
     stream->seekg( offset, std::istream::cur );
-    size_t pos = tell();
-    rewind( 3, 240 );
-    size_t startpos = tell();
-    std::string buffer( pos - startpos, '\0' );
-    stream->read( buffer.data(), pos - startpos );
-    auto it = buffer.begin();
-    for( ; it < buffer.end() && ( *it == '\r' || *it == '\n' ); ++it ) {
-        // skip starting newlines
-    }
-    for( ; it < buffer.end(); ++it ) {
-        if( *it == '\r' ) {
-            err << '\n';
-            if( it + 1 < buffer.end() && *( it + 1 ) == '\n' ) {
-                ++it;
+    // remember positions of several places to print a few lines of context
+    const size_t cursor_pos = tell();               // exact position of error
+    rewind( 1, max_context_chars );                 // rewind to start of line
+    const size_t start_of_line = tell();            // start of error line
+    rewind( max_context_lines, max_context_chars ); // rewind a couple lines to show context
+    const size_t start_of_context = tell();         // start of context (couple lines above)
+    size_t end_of_line;                             // end of error line (without \n)
+
+    {
+        seek( cursor_pos );
+        // find end of line and store in endpos
+        end_of_line = tell();
+        while( stream->peek() != EOF ) {
+            if( stream->peek() == '\n' ) {
+                break;
             }
-        } else {
-            err << *it;
-        }
-    }
-    if( !is_whitespace( peek() ) && stream->good() ) {
-        err << peek();
-    }
-    // display a pointer to the position
-    rewind( 1, 240 );
-    startpos = tell();
-    err << '\n';
-    if( pos > startpos ) {
-        err << std::string( pos - startpos, ' ' );
-    }
-    err << "^\n";
-    seek( pos );
-    // if that wasn't the end of the line, continue underneath pointer
-    char ch = stream->get();
-    if( ch == '\r' ) {
-        if( peek() == '\n' ) {
             stream->get();
+            end_of_line = tell();
         }
-    } else if( ch == '\n' ) {
-        // pass
-    } else if( peek() != '\r' && peek() != '\n' && !stream->eof() ) {
-        for( size_t i = 0; i < pos - startpos + 1; ++i ) {
-            err << ' ';
+    }
+    seek( start_of_context );
+    {
+        // print context lines to start of error line
+        std::string buffer( start_of_line - start_of_context, '\0' );
+        stream->read( buffer.data(), start_of_line - start_of_context );
+        err << buffer;
+    }
+    {
+        // print error line
+        std::string buffer( end_of_line - start_of_line, '\0' );
+        stream->read( buffer.data(), end_of_line - start_of_line );
+        err << color_error << buffer << color_end << "\n";
+    }
+    {
+        // display a cursor at the position if possible, or at start of line
+        err << color_highlight;
+        if( const int padding = cursor_pos - start_of_line - 1; padding > 0 ) {
+            err << std::string( padding, ' ' );
```

Comment:
I believe part of the reason the line was broken into two parts was that, 1. it allows seeing where the error is even if the caret is misplaced, because you won't know the exact relative width of characters in a terminal (or a text editor if the output is redirected) if there are unicode characters in the message, and 2. it allows to easier see where the error is if the text before the caret spans multiple lines.

For 1, `utf8_width` should be used to at least make it work in-game and with monospace font with the same relative width as the in-game code. For 2, I'm not aware of any better solutions than the previous way of breaking the line though.

Reply:
Did the (1) - I had to pull a copy of `UTF8_getch` and `utf8_width` from cata_charset.h/cpp to json.cpp, otherwise it pulled output.h for remove_color_tags() and that pulled string formatting sources... and probably more dependencies, this will need a follow up PR to sort it out

For 2 - afaict the editors and fonts that support monospace "correctly" will work even for CJK glyphs (even though there's close to no CJK or double-width codepoints in json files), even the ones that don't quite work correctly can be made to work using a specific font, for example using Inconsolata ( free font with fixed pitch ) renders CJK correctly everywhere I tested

Reply:
> For 2 - afaict the editors and fonts that support monospace "correctly" will work even for CJK glyphs

There are "ambiguous width" glyphs which can be either 1 or 2 cells wide so it might be difficult to find a font that matches the in-game width exactly for every glyph, but I agree using `utf8_width` is good enough for most cases.